pandas/core/groupby.py

```diff
@@ -4585,9 +4645,17 @@ def _apply_to_column_groupbys(self, func):
              in self._iterate_column_groupbys()),
             keys=self._selected_obj.columns, axis=1)
 
+    def _fill(self, direction, limit=None):
+        """Overriden method to join grouped columns in output"""
+        res = super(DataFrameGroupBy, self)._fill(direction, limit=limit)
+        output = collections.OrderedDict(
+            (grp.name, grp.group_index.take(grp.labels)) for grp in
+            self.grouper.groupings)
+
+        return self._wrap_transformed_output(output).join(res)
```

Comment:
here, rather than joining, could't the entries of `res` be added be added to `output`?  should have the same shape so shouldn't need to go through the join machinery

Reply:
As is the super call already wraps it's `output` into a `DataFrame` or `Series` (actually returned by `_get_cythonized_result`. If we wanted to, we could create some private methods that pass around the "unwrapped" output which private methods in this class and its parent can access before ever calling `_wrap_transformed_output`, but I would question if the juice is really worth the squeeze to do that

Reply:
you can do a concat no?

Reply:
concat should work just as well. Is that out of performance reasons we prefer that to join?

Reply:
just idiomatic